### PR TITLE
fix(aws-amplify/auth): debouncing get user session calls

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

Associated with: https://github.com/aws-amplify/amplify-js-samples-staging/pull/441
Replaces: https://github.com/aws-amplify/amplify-js/pull/10195
Rebasing the fix/auth-debouncing branch to main. 
Code author - Francisco, Allan
Changes allow for debouncing getSession API calls 

#### Description of how you validated changes
unit test
[ineg test](https://app.circleci.com/pipelines/github/aws-amplify/amplify-js/16100/workflows/e23325d9-a8a4-4a79-992c-45b1ccfc104b)

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -507,7 +507,7 @@ jobs:
           framework: vue
           category: auth
           sample_name: authenticator-vue3
-          spec: ui-amplify-authenticator
+          spec: new-ui-authenticator
```
Comment (elorzafe) [inline .circleci/config.yml:510]: this was a change on `samples staging` repo?

File: packages/auth/src/Auth.ts [hunk 2]
```diff
@@ -1629,90 +1629,71 @@ export class AuthClass {
...
-														new Error(
-															`Session is invalid due to: ${err.message} and failed to clean up invalid session: ${cleanUpError.message}`
-														)
-													);
-													return;
-												}
-												rej(err);
-											} else {
-												res(user);
+						const clientMetadata = this._config.clientMetadata; // TODO: verify behavior if this is override during signIn
```
Comment (elorzafe) [inline packages/auth/src/Auth.ts:1643]: Maybe we should verify this

**Review by elorzafe — APPROVED** [review]
Looks good to me

File: .circleci/config.yml (hunk 1 above)
Comment (AllanZhengYP) [reply .circleci/config.yml:510]: It's here: https://github.com/aws-amplify/amplify-js-samples-staging/pull/441

File: packages/auth/src/Auth.ts (hunk 2 above)
Comment (AllanZhengYP) [reply packages/auth/src/Auth.ts:1643]: Good callout!

File: packages/auth/__tests__/auth-unit-test.ts [hunk 3]
```diff
@@ -1814,6 +1814,29 @@ describe('auth unit test', () => {
 			spyon.mockClear();
 		});
 
+		test('debouncer happy case', async () => {
```
Comment (jimblanc) [inline packages/auth/__tests__/auth-unit-test.ts:1817]: Nit: I would be more explicit about what's getting tested here, e.g. 'getSession debouncer`

File: packages/auth/__tests__/auth-unit-test.ts [hunk 4]
```diff
@@ -1814,6 +1814,29 @@ describe('auth unit test', () => {
...
+				Username: 'username',
+				Pool: userPool,
+			});
+
+			expect.assertions(1);
+
+			for (let index = 0; index < 10; index++) {
+				auth.userSession(user);
+			}
+			expect(spyon).toHaveBeenCalledTimes(1);
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/auth/__tests__/auth-unit-test.ts:1835]: The Vue auth test using legacy UI has been flaky before this change. For example, the [test history of datastore laziness branch](https://app.circleci.com/pipelines/github/aws-amplify/amplify-js?branch=datastore-laziness), the particular test fails more than 50% of time even with 3 attempts each execution.

File: packages/auth/src/Auth.ts [hunk 5]
```diff
@@ -1816,47 +1797,63 @@ export class AuthClass {
...
 	}
 
-	/**
-	 * Get the corresponding user session
-	 * @param {Object} user - The CognitoUser object
-	 * @return - A promise resolves to the session
-	 */
-	public userSession(user): Promise<CognitoUserSession> {
+	private async _userSession({
+		user,
```
Comment (jimblanc) [inline packages/auth/src/Auth.ts:1801]: Nit: Does this need to be an object param (vs just doing `_userSession(user: CognitoUser)`)?

**Review by jimblanc — APPROVED** [review]
Looks good to me, few small nits

File: packages/auth/__tests__/auth-unit-test.ts (hunk 3 above)
Comment (AllanZhengYP) [reply packages/auth/__tests__/auth-unit-test.ts:1817]: It unit test is inside `userSession test` test suit, I think it's a bit redundant to prefix it with the test suite name.

File: packages/auth/src/Auth.ts (hunk 2 above)
Comment (AllanZhengYP) [reply packages/auth/src/Auth.ts]: Confirmed the sign in won't override the `Auth._config.clientMetadata`
<Summary>Code Snippet: <Details>

```javascript
import { Auth } from "@aws-amplify/auth";
import { useState } from "react";
import awsconfig from "./aws-exports";

Auth.configure(awsconfig);
Auth.configure({
  clientMetadata: {
    test: 'default metadata'
  }
})

function App() {
  const [user, setUser] = useState(null);

  async function signIn() {
    const [user1, user2] = await Promise.all([
      Auth.signIn({
        username: 'test01',
        password: 'The#test1'
      }, null, {
        test: 'some metadat'
      }),
      Auth.signIn({
        username: 'test01',
        password: 'The#test1'
      }, null, {
        test: 'another metadata'
      }),
    ]);
    setUser(user2);
  }

  async function currentUser() {
    const user = await Auth.currentAuthenticatedUser({ bypassCache: true });
    setUser(user);
  }

  async function userSession() {
    await Auth.userSession(user);
  }
  return (
    <div className="App">
      <header className="App-header">
        <button onClick={signIn}>sign in</button>
        <button onClick={currentUser}>refresh token</button>
        <button onClick={userSession}>user session</button>
      </header>
    </div>
  );
}

export default App;

```

</Details></Summary>

**Review by elorzafe — APPROVED** [review]
Looks good, thanks @AllanZhengYP 🎉

**Pull request merged**